zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/remote/RemoteInterpreterDownloader.java

```diff
@@ -0,0 +1,148 @@
+    if (args.length > 0) {
+      zeppelinServerHost = args[0];
+      port = Integer.parseInt(args[1]);
+      interpreter = args[2];
+      localRepoPath = args[3];
+    }
+    RemoteInterpreterEventClient intpEventClient = new RemoteInterpreterEventClient(
+        zeppelinServerHost, port, 3);
+
+    RemoteInterpreterDownloader downloader = new RemoteInterpreterDownloader(interpreter,
+        intpEventClient, new File(localRepoPath));
+    downloader.syncAllLibraries();
+  }
+
+  private void syncAllLibraries() {
+    LOGGER.info("Loading all libraries for interpreter {} to {}", interpreter, localRepoDir);
+    List<LibraryMetadata> metadatas = client.getAllLibraryMetadatas(interpreter);
+    if (!localRepoDir.isDirectory()) {
+      LOGGER.error("{} is no directory", localRepoDir);
+      return;
+    }
+    Set<String> syncedLibraries = new HashSet<>();
+    // Add or update new libraries
+    for (LibraryMetadata metadata : metadatas) {
+      File library = new File(localRepoDir, metadata.getName());
+      addOrUpdateLibrary(library, metadata);
+      syncedLibraries.add(metadata.getName());
+    }
+    // Delete old Jar files
+    for (File file : FileUtils.listFiles(localRepoDir, new String[] { "jar" }, false)) {
+      if (!syncedLibraries.contains(file.getName())) {
+        try {
+          LOGGER.info("Delete {}, because it's not present on the server side", file.toPath());
+          Files.delete(file.toPath());
+        } catch (IOException e) {
+          LOGGER.error("Unable to delete old library {} during sync.", file, e);
+        }
+      }
+    }
+  }
+
+  private void addOrUpdateLibrary(File library, LibraryMetadata metadata) {
+    try {
+      if (library.exists() && library.canRead()) {
+        long localChecksum = FileUtils.checksumCRC32(library);
+        if (localChecksum == metadata.getChecksum()) {
+          // nothing to do if checksum is matching
+          LOGGER.info("Library {} is present ", library.getName());
+        } else {
+          // checksum didn't match
+          Files.delete(library.toPath());
+          downloadLibrary(library, metadata);
+        }
+      } else {
+        downloadLibrary(library, metadata);
+      }
+    } catch (IOException e) {
+      LOGGER.error("Can not add or update library {}", library, e);
+    }
+  }
+
+  private void downloadLibrary(File library, LibraryMetadata metadata) {
+    LOGGER.debug("Trying to download library {} to {}", metadata.getName(), library);
+    try {
+      if (library.createNewFile()) {
+        int attempt = 0;
+        while (attempt < MAX_LIBRARY_DOWNLOAD_ATTEMPTS) {
+          ByteBuffer bytes = client.getLibrary(interpreter, metadata.getName());
+          if (bytes == null) {
+            LOGGER.error("ByteBuffer of library {} is null."
+                + " For a detailed message take a look into Zeppelin-Server log. Attempt {} of {}",
+                metadata.getName(), ++attempt, MAX_LIBRARY_DOWNLOAD_ATTEMPTS);
```

Comment:
Do you think the error if the function 'getLibrary()' returns null can be recovered by redoing it?

Reply:
In principle, this is a network communication. We already handle ThriftExceptions in the `PooledRemoteClient`, so these should not occur.
I think a retry is not bad because IOExceptions can occur in Zeppelin Server. What is your opinion on this?
Btw. The happy path never returns null.

Reply:
It doesn't matter if you try a few more attempts.